v1.1/parallel-statement-execution.md

```diff
@@ -0,0 +1,128 @@
+
+Traditional SQL engines execute the statements sequentially without affecting performance. But in case of globally-distributed CockroachDB clusters, sequential execution of statements adds to aggregate latency. This is because to maintain strong consistency across a globally-distributed cluster, CockroachDB replicates writes using the Raft [consensus protocol](https://www.cockroachlabs.com/blog/consensus-made-thrive/). The protocol requires majority of replicas to agree before a write is committed. Replicas always live on separate nodes, and communication between these nodes takes time. Executing SQL statements sequentially leads to higher cumulative communication latency. 
+
+With parallel statement execution, however, multiple SQL statements within a transaction are executed at the same time, thereby reducing the aggregate latency over multiple SQL statements. 
+
+## How Parallel Statement Execution Works
+
+SQL statements within a single transaction can be executed in parallel if the statements are independent. CockroachDB considers SQL statements within a single transaction to be independent if their execution can be safely reordered without affecting their results. 
+
+As seen earlier, sequential execution of SQL statements returns a return value immediately from the server. The next statement is executed only after the return value is received. Appending the `RETURNING NOTHING` clause with SQL statements prevents the server from sending return values and allows CockroachDB to execute the statements in parallel. To execute statements in parallel, append the `RETURNING NOTHING` suffix to the SQL statements. The statements are executed in parallel till CockroachDB encounters a **barrier statement**. A barrier statement is a statement that needs to be executed sequentially.
+
+In our example, the transaction would be as follows:
+
+~~~ sql
+> BEGIN;
+> UPDATE username SET last_name = 'Smith' WHERE id = 1 RETURNING NOTHING;
+> UPDATE movies SET favorite_movie = 'The Matrix' WHERE id = 1 RETURNING NOTHING;
+> UPDATE songs SET favorite_song = 'All this time' WHERE id = 1 RETURNING NOTHING;
+> COMMIT;
+~~~
+
+In this case, because the statements within the transaction are independent of each other, they can be executed in parallel without affecting the results. The COMMIT statement is executed sequentially, and thus acts as a barrier statement. Barrier statements are executed only after all the parallel statements are executed.
+
+The following conceptual diagram shows how the transaction is executed sequentially and in parallel. The diagram also shows how executing statements in parallel reduces the aggregate latency.
+
+<img src="{{ 'images/Parallel_Statement_Normal_Execution.png' | relative_url }}" alt="CockroachDB Parallel Statement Execution" style="border:1px solid #eee;max-width:100%" />
```

Comment:
Do we need the sequential timeline again? It's right above this diagram already.

Reply:
Side note, the diagrams are great!! Nice job 🎉

Reply:
I like the side-by-side comparison diagram that shows the performance improvement. Hence the repetition :)

Reply:
Ah yeah, that's a very good point.